src/ipa_pc/mod.rs

```diff
@@ -1814,19 +1814,69 @@ mod tests {
         use algebra::fields::tweedle::fr::Fr;
         use algebra::UniformRand;
 
-        let seed = b"";
-        let mut rng1 = FiatShamirChaChaRng::<Blake2s>::from_seed(seed);
-        let mut rng2 = FiatShamirChaChaRng::<Blake2s>::from_seed(seed);
+        // Empty test
+        {
+            let mut rng1 = FiatShamirChaChaRng::<Blake2s>::from_seed(FiatShamirChaChaRngSeed::default().finalize());
+            let mut rng2 = FiatShamirChaChaRng::<Blake2s>::default();
+
+            assert_eq!(rng1.get_state(), rng2.get_state());
+
+            let a = Fr::rand(&mut rng1);
+            let b = Fr::rand(&mut rng2);
+
+            assert_eq!(a, b);
+            assert_eq!(rng1.get_state(), rng2.get_state());
+
+            rng1.absorb(b"ABSORBABLE_ELEM");
+            rng2.absorb(b"ABSORBABLE_ELEM");
 
-        let a = Fr::rand(&mut rng1);
-        let b = Fr::rand(&mut rng2);
+            assert_eq!(rng1.get_state(), rng2.get_state());
 
-        assert_eq!(a, b);
+            let a: Fr = rng1.squeeze_128_bits_challenge();
+            let b: Fr = rng2.squeeze_128_bits_challenge();
+
+            assert_eq!(a, b);
+            assert_eq!(rng1.get_state(), rng2.get_state());
+        }
+
+        // No cross protocol attacks possible
+        {
+            let fs_rng_seed = {
+                let mut seed_builder = FiatShamirChaChaRngSeed::new();
+                seed_builder.add_bytes(b"TEST_SEED").unwrap();
+                seed_builder.finalize()
+            };
 
-        let a: Fr = rng1.squeeze_128_bits_challenge();
-        let b: Fr = rng2.squeeze_128_bits_challenge();
+            let malicious_fs_rng_seed = {
+                let mut seed_builder = FiatShamirChaChaRngSeed::new();
+                seed_builder.add_bytes(b"TEST_").unwrap();
+                seed_builder.add_bytes(b"SEED").unwrap();
+                seed_builder.finalize()
+            };
+
+            let mut fs_rng = FiatShamirChaChaRng::<Blake2s>::from_seed(fs_rng_seed);
+            let mut malicious_fs_rng = FiatShamirChaChaRng::<Blake2s>::from_seed(malicious_fs_rng_seed);
+
+            assert_ne!(fs_rng.get_state(), malicious_fs_rng.get_state());
+
+            let a = Fr::rand(&mut fs_rng);
+            let b = Fr::rand(&mut malicious_fs_rng);
+
+            assert_ne!(a, b);
+            assert_ne!(fs_rng.get_state(), malicious_fs_rng.get_state());
+
+            fs_rng.absorb(b"ABSORBABLE_ELEM");
+            malicious_fs_rng.absorb(b"ABSORBABLE_ELEM");
+
+            assert_ne!(fs_rng.get_state(), malicious_fs_rng.get_state());
+
+            let a: Fr = fs_rng.squeeze_128_bits_challenge();
+            let b: Fr = malicious_fs_rng.squeeze_128_bits_challenge();
+
+            assert_ne!(a, b);
+            assert_ne!(fs_rng.get_state(), malicious_fs_rng.get_state());
+        }
```

Comment:
It is not wrong to have an extra test for that, but somehow artificial.
In case we decide to implement ``CanonicalSerialize`` for our seed, this test can be replaced by showing that the deserialization inverts serialization.